Enhance Docker

oliviermichaelis opened the pull request:
* Refactored deprecated functionalities:
  - Used 'celery worker' command instead of 'superset worker' which is
    deprecated since 0.26.0
  - Used 'gunicorn' command instead of 'superset runserver'
* ~~Changes made to allow a dynamic port configuration of superset at runtime rather
than at buildtime~~
  - ~~Port is now dynamically exposed in docker-compose (default is 8088)~~
  - ~~Moved docker healthcheck from Dockerfile to docker-compose~~

mistercrunch commented:
@victornoel do you mind taking a look since you know most above the current docker setup?

victornoel commented on contrib/docker/.env:
```diff
@@ -1 +1,2 @@
 COMPOSE_PROJECT_NAME=superset
+SUPERSET_PORT=8088
```
@oliviermichaelis I don't think it is a good idea to put the env var here: it should reside inside the docker-compose file, so that it is visible to users.
The reason there is an `.env` file is only because it is the only way to configure a project name in a persistent way.

oliviermichaelis replied:
@victornoel I totally agree with you!

victornoel commented on contrib/docker/docker-compose.yml:
```diff
@@ -31,11 +35,12 @@ services:
       SUPERSET_ENV: development
     ports:
-      - 8088:8088
+      - ${SUPERSET_PORT}:${SUPERSET_PORT}
```
@oliviermichaelis you don't have to use the same port for in-container process and host system: I think it would be a better idea to have a fixed port in the Dockerfile as before and if the user wants to change the port, it can change it here on left side of this line. See https://docs.docker.com/compose/compose-file/#ports

oliviermichaelis replied:
@victornoel you are right, I wasn't thinking about the possibility to map a different host port to the container. Thanks for the feedback! :)

victornoel commented on contrib/docker/docker-entrypoint.sh:
```diff
@@ -4,14 +4,18 @@ set -ex
     exec "$@"
 elif [ "$SUPERSET_ENV" = "development" ]; then
-    superset worker &
+    celery worker --app=superset.sql_lab:celery_app --pool=gevent -Ofair &
```
@mistercrunch can you validate it is the correct command? I have no idea :)

mistercrunch replied:
command looks good

victornoel commented on contrib/docker/docker-entrypoint.sh:
```diff
@@ -4,14 +4,18 @@ set -ex
     (cd superset/assets/ && yarn && yarn run sync-backend)
     (cd superset/assets/ && yarn run dev) &
-    flask run -p 8088 --with-threads --reload --debugger --host=0.0.0.0
+    FLASK_APP=superset:app flask run -p ${SUPERSET_PORT} --with-threads --reload --debugger --host=0.0.0.0
```
@mistercrunch can you validate it is the correct command? I have no idea :)

mistercrunch replied:
command looks good

victornoel commented on contrib/docker/docker-entrypoint.sh:
```diff
@@ -4,14 +4,18 @@ set -ex
+        --timeout 60 -b  0.0.0.0:${SUPERSET_PORT} \
+        --limit-request-line 0 \
+        --limit-request-field_size 0 \
+        superset:app
```
@mistercrunch can you validate those are the correct commands? I have no idea :)

mistercrunch replied:
command looks good

victornoel reviewed (changes requested):
@oliviermichaelis basically I think the port change is useless because docker-compose and docker are made so that you don't need to do this. You can just use a fixed port in the container and use docker-compose `ports` configuration to decide how to expose it. The rest I think @mistercrunch should validate them :)

oliviermichaelis commented:
As explained by @victornoel my changes to change the port weren't necessary. I'll remove them.

@victornoel what do you think about tagging the superset image in `docker-compose.yml` with:
```
image: apache/incubator-superset:latest
```

victornoel commented:
> As explained by @victornoel my changes to change the port weren't necessary. I'll remove them.
> 
> @victornoel what do you think about tagging the superset image in `docker-compose.yml` with:
> 
> ```
> image: apache/incubator-superset:latest
> ```

Maybe it's not desired, because the image built is a very dev-oriented image with extra stuffs. But If I wanted to tag it, I would use a simple name (because the `/` separated name implies it's on dockerhub?) and I would use the tag dev, debug or something like that.

Agood question to ask in order to know how to tag is: what would be the usage of such image outside of this docker-compoe?

phoenix24 commented:
@victornoel as part of the pull-request, do these need to be updated as well?  
- `install/helm/superset/values.yaml` 
- `install/helm/superset/templates/deployment.yaml`

victornoel commented:
@phoenix24 very good question, I didn't know those existed :D 
I don't think so because the helm chart rely on @amancevice image, not the one used here (which is dedicated to development, not production).

mistercrunch commented:
LGTM oh my side @victornoel @oliviermichaelis ready for me to merge?

victornoel commented:
@mistercrunch I don't think @oliviermichaelis applied the changes we discussed (or didn't push them because they are marked as resolved)

oliviermichaelis commented:
@mistercrunch @victornoel sorry I didn't have time yet to apply and push the changes yet. I'll try to do that as soon as possible!

kakoni commented:
So now that we are enhancing, is exposing redis and psql ports a needed feature? (6379:6379 and 5432:5432)

amancevice commented:
No, I don’t think so. You would want those apps to run in separate containers.

victornoel commented:
I don't think exposing ports to the host has anything to do with separating
services in containers.

Currently they are in separate containers and see each other through the
private network created by docker-compose.

If the ports are exposed to the host, it can be helpful for debugging
purpose. This compose file and docker image are here for development, so it
makes sense to expose the ports.

Le sam. 22 déc. 2018 à 22:06, Alexander Mancevice <notifications@github.com>
a écrit :

> No, I don’t think so. You would want those apps to run in separate
> containers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-superset/pull/6504#issuecomment-449597588>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJ0z5LjqYqQMNh60fXEYyL-WTppzEdJks5u7p7IgaJpZM4ZJndm>
> .
>

kakoni commented:
@victornoel Got it, it does make sense to expose.

amancevice commented:
Ah, I misunderstood the intent/conflated GitHub issues. My mistake.

oliviermichaelis committed:
Enhance Docker

Refactored deprecated functionalities:
  - Used 'celery worker' command instead of 'superset worker' which is
    deprecated since 0.26.0
  - Used 'gunicorn' command instead of 'superset runserver'

oliviermichaelis committed:
Merge branch 'master' into enhance_docker

oliviermichaelis commented:
I resolved the merge conflicts with the last commits

victornoel reviewed (approved):
@mistercrunch seems good to me

oliviermichaelis commented:
Thanks! :)
